REGRESSION (iOS 17): PWA startup "FetchEvent.respondWith received an error TypeError: Internal error"

youennf opened the pull request:
#### dd68eb88a64ca414570729a171576e6a5571b5c4
<pre>
REGRESSION (iOS 17): PWA startup &quot;FetchEvent.respondWith received an error TypeError: Internal error&quot;
<a href="https://bugs.webkit.org/show_bug.cgi?id=261767">https://bugs.webkit.org/show_bug.cgi?id=261767</a>
rdar://115747046

Reviewed by Chris Dumez and Sihui Liu.

We were skipping sending reference signals when creating caches previously created in other web pages.
This was allowing network process to clean up memory and we ended up not matching any resource from DOMCache.

Make sure to always send the reference message, but only send dereference message if there was a reference message before.

In addition, there is a potential race when network process sends DOMCache identifiers to the web process, but web process did not yet created the DOMCache objects
(aka WebProcess has not sent the reference message).
In that case, the OriginStorageManager can be removed, thus making the DOMCache identifiers useless.
To prevent this, we are now locking/unlocking cache storage from WebProcess when creating DOMCache objects.
This will prevent OriginStorageManager to be removed.

We are also handling reference/dereference counters directly in WebProcess since the new code path in networking process is not doing so anymore.
This is needed as the same DOMCache identifier may be used by different DOMCache objects (say one in a service worker and one in window).

Covered by API test.

* Source/WebCore/Modules/cache/CacheStorageConnection.h:
* Source/WebCore/Modules/cache/DOMCacheStorage.cpp:
(WebCore::ConnectionStorageLock::ConnectionStorageLock):
(WebCore::ConnectionStorageLock::~ConnectionStorageLock):
(WebCore::DOMCacheStorage::retrieveCaches):
* Source/WebCore/Modules/cache/WorkerCacheStorageConnection.cpp:
(WebCore::WorkerCacheStorageConnection::lockStorage):
(WebCore::WorkerCacheStorageConnection::unlockStorage):
* Source/WebCore/Modules/cache/WorkerCacheStorageConnection.h:
* Source/WebCore/page/CacheStorageProvider.h:
* Source/WebKit/NetworkProcess/storage/CacheStorageManager.cpp:
(WebKit::CacheStorageManager::lockStorage):
(WebKit::CacheStorageManager::unlockStorage):
(WebKit::CacheStorageManager::connectionClosed):
(WebKit::CacheStorageManager::isActive):
* Source/WebKit/NetworkProcess/storage/CacheStorageManager.h:
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::lockCacheStorage):
(WebKit::NetworkStorageManager::unlockCacheStorage):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.messages.in:
* Source/WebKit/WebProcess/Cache/WebCacheStorageConnection.cpp:
(WebKit::WebCacheStorageConnection::open):
(WebKit::WebCacheStorageConnection::reference):
(WebKit::WebCacheStorageConnection::dereference):
(WebKit::WebCacheStorageConnection::lockStorage):
(WebKit::WebCacheStorageConnection::unlockStorage):
(WebKit::WebCacheStorageConnection::networkProcessConnectionClosed):
* Source/WebKit/WebProcess/Cache/WebCacheStorageConnection.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ServiceWorkerBasic.mm:

Canonical link: <a href="https://commits.webkit.org/268565@main">https://commits.webkit.org/268565@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fcc5111be62315d1165adfca14ca132c8cf0e7e0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/19910 "4 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/20334 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/20954 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/21803 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/18596 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/20146 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/23591 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/20483 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/20119 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/20132 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/20100 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/17319 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/22657 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/17271 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/18114 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/24381 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/18347 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/18290 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/22367 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/18881 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/16007 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/18052 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4807 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/22401 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/18705 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/WebProcess/Cache/WebCacheStorageConnection.cpp:
```diff
@@ -92,18 +85,34 @@ void WebCacheStorageConnection::batchPutOperation(WebCore::DOMCacheIdentifier ca
-    if (!m_connectedIdentifiers.contains(cacheIdentifier))
-        return;
-
+    m_connectedIdentifiers.add(cacheIdentifier);
```
Why don't we early return if we already have this identifier in the set?

youennf replied:
The same DOMCache may be created several times in the same WebProcess (different contexts).
The networking process is supposed to count these instances.
But the new code path is not doing so anymore so we might best do this in WebProcess.
Or maybe we should stop using identifiers and pass ClientOrigin + cache name for every request to better manage network process crashes.

youennf replied:
I updated to use counters in WebProcess.

cdumez commented on Source/WebKit/WebProcess/Cache/WebCacheStorageConnection.h:
```diff
@@ -63,13 +63,16 @@ class WebCacheStorageConnection final : public CanMakeWeakPtr<WebCacheStorageCon
 
     WebCacheStorageProvider& m_provider;
     HashSet<WebCore::DOMCacheIdentifier> m_connectedIdentifiers;
+    HashMap<ClientOrigin, uint64_t> m_clientOriginLockRequestCounters;
```
Seems like a use case for HashCountedSet?

youennf replied:
Will do

szewai commented on Source/WebCore/Modules/cache/WorkerCacheStorageConnection.cpp:
```diff
@@ -229,6 +231,20 @@ void WorkerCacheStorageConnection::dereference(DOMCacheIdentifier cacheIdentifie
 
+void WorkerCacheStorageConnection::lockStorage(const ClientOrigin& origin)
+{
+    callOnMainThread([mainThreadConnection = m_mainThreadConnection, origin = origin.isolatedCopy()]() {
```
It seems there could still be race if LockCacheStorage is a separate message from CacheStorageAllCaches (e.g. network process might clear in-memory storage after it receives CacheStorageAllCaches and before LockCacheStorage). 
How about making network process automatically ref all caches for the web process connection when it receives CacheStorageAllCaches, and make web process send a DidRetrieveAllCaches/DereferenceCaches message to tell network process to remove ref to all caches (or caches it does not need to connect) in CacheStorageAllCaches reply? Web process could skip the cache identifier that is used to create new DOMCache.

This might be applied to CacheStorageOpenCache: network process ref the cache when receiving the message, and web process should send CancelOpenCache/DereferenceCache to tell network process to deref the cache when it decides to not create a new DOMCache with identifier or the DOMCache is destroyed.

youennf replied:
LockCacheStorage message happens before CacheStorageAllCaches message so there should be no race.
I thought about doing your approach but it seemed to me less readable (we would need to pass a callback to be executed on either worker or main thread as part of the CacheStorageAllCaches result callback).

szewai replied:
ok, then let's go with your implementation then

szewai reviewed (approved):
r=me

youennf committed:
REGRESSION (iOS 17): PWA startup "FetchEvent.respondWith received an error TypeError: Internal error"
https://bugs.webkit.org/show_bug.cgi?id=261767
rdar://115747046

Reviewed by Chris Dumez and Sihui Liu.

We were skipping sending reference signals when creating caches previously created in other web pages.
This was allowing network process to clean up memory and we ended up not matching any resource from DOMCache.

Make sure to always send the reference message, but only send dereference message if there was a reference message before.

In addition, there is a potential race when network process sends DOMCache identifiers to the web process, but web process did not yet created the DOMCache objects
(aka WebProcess has not sent the reference message).
In that case, the OriginStorageManager can be removed, thus making the DOMCache identifiers useless.
To prevent this, we are now locking/unlocking cache storage from WebProcess when creating DOMCache objects.
This will prevent OriginStorageManager to be removed.

We are also handling reference/dereference counters directly in WebProcess since the new code path in networking process is not doing so anymore.
This is needed as the same DOMCache identifier may be used by different DOMCache objects (say one in a service worker and one in window).

Covered by API test.

* Source/WebCore/Modules/cache/CacheStorageConnection.h:
* Source/WebCore/Modules/cache/DOMCacheStorage.cpp:
(WebCore::ConnectionStorageLock::ConnectionStorageLock):
(WebCore::ConnectionStorageLock::~ConnectionStorageLock):
(WebCore::DOMCacheStorage::retrieveCaches):
* Source/WebCore/Modules/cache/WorkerCacheStorageConnection.cpp:
(WebCore::WorkerCacheStorageConnection::lockStorage):
(WebCore::WorkerCacheStorageConnection::unlockStorage):
* Source/WebCore/Modules/cache/WorkerCacheStorageConnection.h:
* Source/WebCore/page/CacheStorageProvider.h:
* Source/WebKit/NetworkProcess/storage/CacheStorageManager.cpp:
(WebKit::CacheStorageManager::lockStorage):
(WebKit::CacheStorageManager::unlockStorage):
(WebKit::CacheStorageManager::connectionClosed):
(WebKit::CacheStorageManager::isActive):
* Source/WebKit/NetworkProcess/storage/CacheStorageManager.h:
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::lockCacheStorage):
(WebKit::NetworkStorageManager::unlockCacheStorage):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.messages.in:
* Source/WebKit/WebProcess/Cache/WebCacheStorageConnection.cpp:
(WebKit::WebCacheStorageConnection::open):
(WebKit::WebCacheStorageConnection::reference):
(WebKit::WebCacheStorageConnection::dereference):
(WebKit::WebCacheStorageConnection::lockStorage):
(WebKit::WebCacheStorageConnection::unlockStorage):
(WebKit::WebCacheStorageConnection::networkProcessConnectionClosed):
* Source/WebKit/WebProcess/Cache/WebCacheStorageConnection.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ServiceWorkerBasic.mm:

Canonical link: https://commits.webkit.org/268565@main
